### PR TITLE
refactor: allOf interpretation

### PR DESCRIPTION
**Pull request by jonaslagoni**
**Description**
This PR adds the refactoring of `allOf` to the new interpreter.

It changes the entire perspective, to instead of adding simplification functions per CommonModel property, we process the JSON Schema keywords which then affects the CommonModel in some way.

**It adds**
- InterpretAllOf, to interpret the `allOf` schema keyword.

**Files this PR is supposed to replace**
- https://github.com/asyncapi/generator-model-sdk/blob/master/src/simplification/SimplifyExtend.ts
- And all related tests have been refactored to have next to no dependencies by using the jest mocking framework.

**Todo**

- [x] Refactor docs
- [x] Solve code smells
- [x] Add logging to the process
- [x] Inspect tests
- [x] Rename to transformer

**Related issue(s)**
Blocked by #184

### CODE REVIEW

File: src/interpreter/InterpretAllOf.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * It either merges allOf schemas into existing model or if allowed, create inheritance.
+ * 
+ * @param schema 
+ * @param model 
+ * @param interpreter 
+ */
+export default function interpretAllOf(schema: Schema | boolean, model: CommonModel, interpreter : Interpreter) {
+  if (typeof schema === 'boolean' || schema.allOf === undefined) return;
+
+  for (const allOfSchema of (schema.allOf)) {
```
Comment (magicmatatjahu) [inline src/interpreter/InterpretAllOf.ts:19]: Why `(schema.allOf)` in brackets? 😄

File: src/interpreter/InterpretAllOf.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  for (const allOfSchema of (schema.allOf)) {  
+    const interpretedModels = interpreter.interpret(allOfSchema);
+    if (interpretedModels.length === 0) continue;
+    const interpretedModel = interpretedModels[0];
+    if (isModelObject(interpretedModel) && interpreter.options.allowInheritance === true) {
+      Logger.info(`Processing allOf, inheritance is allowed, ${model.$id} inherits from ${interpretedModel.$id}`, model, interpretedModel);
+      model.addExtendedModel(interpretedModel);
+    } else {
+      Logger.info('Processing allOf, inheritance is not allowed. AllOf model is merged together with already interpreted model', model, interpretedModel);
```
Comment (magicmatatjahu) [inline src/interpreter/InterpretAllOf.ts:27]:
```suggestion
      Logger.info('Processing allOf, inheritance is enabled. AllOf model is merged together with already interpreted model', model, interpretedModel);
```

File: src/models/CommonModel.ts [hunk 3]
```diff
@@ -100,6 +100,27 @@ export class CommonModel extends CommonSchema<CommonModel> {
...
+   * @param extendedModel 
+   */
+  addExtendedModel(extendedModel: CommonModel) {
+    if (extendedModel.$id === undefined) {
+      Logger.error('Found no $id for allOf model and cannot extend the existing model, this should never happen.', this, extendedModel);
+      return;
+    }
+    if (this.extend === undefined) {
+      this.extend = [];
+    } else if (this.extend.includes(extendedModel.$id)) {
```
Comment (magicmatatjahu) [inline src/models/CommonModel.ts:117]:
```suggestion
    if (extendedModel.$id === undefined) {
      Logger.error('Found no $id for allOf model and cannot extend the existing model, this should never happen.', this, extendedModel);
      return;
    }
    this.extend = this.extend || [];
    if (this.extend.includes(extendedModel.$id)) { 
```

File: src/interpreter/InterpretAllOf.ts (hunk 1 above)
Comment (jonaslagoni) [reply src/interpreter/InterpretAllOf.ts:19]: Yea that is an excellent question 🤔

Comment (jonaslagoni) [reply src/interpreter/InterpretAllOf.ts]: removed

File: src/interpreter/InterpretAllOf.ts (hunk 2 above)
Comment (jonaslagoni) [reply src/interpreter/InterpretAllOf.ts:27]: it should be `not allowed` or `not enabled`, since it depends on the interpretation options.

Comment (magicmatatjahu) [reply src/interpreter/InterpretAllOf.ts:27]: Yeah, sorry 🤦🏼 I thought about `not enabled`... Sorry!

Comment (jonaslagoni) [reply src/interpreter/InterpretAllOf.ts:27]: I am gonna change `inheritance is allowed` to `inheritance is enabled` to be consistant.

**Review by magicmatatjahu — APPROVED** [review]
LGTM! :)

**Comment by jonaslagoni** [comment]
Thanks for the review 👍

**Pull request merged**